cmd/release-controller/http_candidate.go

```diff
@@ -0,0 +1,452 @@
+		found := false
+		for _, stream := range releaseStreams {
+			streamVersion, _ := semverParseTolerant(stream)
+			if next[stream] == nil && streamVersion.Major == v.Major && streamVersion.Minor == v.Minor {
+				found = true
+				break
+			}
+		}
+		if !found {
+			continue
+		}
+
+		// Call oc adm release info to get previous nightly info for the stable release
+		op, err := c.releaseInfo.ReleaseInfo(r.from)
+		if err != nil {
+			// releaseinfo not found, old tag
+			return releaseCandidates, fmt.Errorf("Could not get release info for tag %s: %v", r.from, err)
+		}
+
+		type releaseInfoShort struct {
+			Image      string               `json:"image"`
+			References *imagev1.ImageStream `json:"references"`
+		}
+
+		releaseInfo := releaseInfoShort{}
+		if err := json.Unmarshal([]byte(op), &releaseInfo); err != nil {
+			return releaseCandidates, fmt.Errorf("Could not unmarshal release info for tag %s: %v", r.from, err)
+		}
+		latestPromotedFrom := releaseInfo.References.Annotations[releaseAnnotationFromRelease]
+		// latestPromotedFrom has the format <registry>:<previous nightly> or <previous nightly>
+		// e.g: registry.svc.ci.openshift.org/ocp/release:4.1.0-0.nightly-2019-07-18-192922
+		// Removing prefix
+		if idx := strings.LastIndex(latestPromotedFrom, ":"); idx != -1 {
+			latestPromotedFrom = latestPromotedFrom[idx+1:]
+		}
+
+		// Find the creation time and the stream for the nightly this stable release was promoted from
+		var timeString, stream string
+		prevTags, _ := c.findReleaseStreamTags(false, latestPromotedFrom)
+		if prevTags[latestPromotedFrom] != nil &&
+			prevTags[latestPromotedFrom].Tag.Annotations[releaseAnnotationCreationTimestamp] != "" {
+			// Use previous release stream tags, if available
+			stream = prevTags[latestPromotedFrom].Tag.Annotations[releaseAnnotationName]
+			timeString = prevTags[latestPromotedFrom].Tag.Annotations[releaseAnnotationCreationTimestamp]
+		} else if rePreviousReleaseName.MatchString(latestPromotedFrom) {
+			//timestamps from elsewhere
+			// Try to use name format of previous nightly to find release stream and timestamp
+			stream = rePreviousReleaseName.ReplaceAllString(latestPromotedFrom, "${STREAM}")
+			// Use creation time of latest stable release instead
+			timeString = releaseInfo.References.CreationTimestamp.Format(time.RFC3339)
+		} else {
+			glog.Errorf("Could not find tag %s , tag may have been deleted", latestPromotedFrom)
+			continue
+		}
+
+		// Check if selected stream belongs to any we are interested in
+		found = false
+		for _, s := range releaseStreams {
+			if stream == s {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// The stable release belongs to a release stream we are not processing
+			continue
+		}
+
+		pt, err := time.Parse(time.RFC3339, timeString)
+		if err != nil {
+			glog.Errorf("Unable to parse timestamp %s for %s: %v", timeString, latestPromotedFrom, err)
+			continue
+		}
+		remaining--
+		latestPromotedTime[stream] = pt.Unix()
```

Comment:
Let's draw out the algorithm here.  

Goal: find the most recently promoted tag for a stream (out of nightly, but could be anything in the future)

1. Identify the newest (according to semantic sort) stable tag in the same major.minor version from the stable streams
2. Retrieve what it was promoted from - if it wasn't promoted from anything, we don't know what the latest is, so we should probably stop here and return nothing to the user
3. Once we have what it was promoted from, we want to find it in the set of tags we currently have for the stream and attempt to filter the list by semantic tags newer than the semantic tag of the release
4. If the current tag does not exist exactly, we want to infer, based on the tag as a semantic version, which tags are semantically newer (odds are almost 100% it's older than all the others)
5. if we can't make that inference (the tag isn't semantic) we should return an error
6. once we have the list of tags newer semantically than the last promoted, we want to iterate down to find the first accepted (today) and/or highest test pass rate (future) and/or the one with the keep annotation in a special form (future)

So I think you can simplify this logic quite a bit - I would recommend trying to break this into chunks / sub functions (if that helps simplify the flow) into the different steps, and each step reduce the inputs until you have a list, exiting early if you don't.  For instance, you may want to accumulate all the matching candidate tags in your stream first, then keep referencing / pruning that list down.